clippy_lints/src/methods/redundant_repeat_take.rs

```diff
@@ -0,0 +1,32 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::LateContext;
+
+use super::REDUNDANT_REPEAT_TAKE;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, repeat_expr: &Expr<'_>, take_arg: &Expr<'_>) {
+    if !expr.span.from_expansion()
+        && let ExprKind::Call(inner, repeat_args) = repeat_expr.kind
+        && repeat_args.len() == 1
+        && let ExprKind::Path(ref qpath) = inner.kind
+        && let QPath::Resolved(_, path) = qpath
+        && path.segments.len() == 1
+        && path.segments[0].ident.name.as_str() == "repeat"
+    {
+        span_lint_and_sugg(
+            cx,
+            REDUNDANT_REPEAT_TAKE,
+            expr.span,
+            "redundant `repeat().take()`",
+            "consider using `repeat_n()` instead",
+            format!(
+                "std::iter::repeat_n({}, {})",
+                snippet(cx, repeat_args[0].span, ""),
+                snippet(cx, take_arg.span, "")
+            ),
+            Applicability::MachineApplicable,
```

Comment:
FWIW, this changes the expression's type so it can lead to compile errors in some cases where its type is unified with other branches, e.g.:
```rs
// From an external library the user doesn't get to control
mod external_lib {
  pub fn iter() -> std::iter::Take<std::iter::Repeat<&'static [u8]>> { todo!() }
}

let iter = match 1 {
  1 => external_lib::iter(),
  2 => std::iter::repeat([1,2].as_slice()).take(2) // can't really change this to repeat_n
};
```
so in those situations `MachineApplicable` could lead to issues. There have been a few false reports of the like in the past (#11761 comes to mind).

The [`is_expr_used_or_unified`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_expr_used_or_unified.html) util function could help avoid these I think (not emitting a warning if that returns true).